### PR TITLE
add mingw support to mysql

### PR DESCRIPTION
**Pull request by heheda123123**

### CODE REVIEW

**Comment by heheda123123** [comment]
编译运行都没问题 ，但是为啥xmake编译之后，不会自动把dll复制到exe目录下去呢

File: packages/m/mysql/xmake.lua [hunk 1]
```diff
@@ -36,6 +43,12 @@ package("mysql")
         end
     end)
 
+    on_install("mingw", function (package)
+        os.cp("mysql/8.0.31/ad611e3230894c35896b2ccae5be9c7e/lib", package:installdir())
```
Comment (waruqi) [inline packages/m/mysql/xmake.lua:47]: 不要硬编码版本，从 `package:version_str()` 取下，或者改成 `mysql/*/*/lib` 更通用些

File: packages/m/mysql/xmake.lua [hunk 2]
```diff
@@ -16,6 +16,13 @@ package("mysql")
         add_configs("shared", {description = "Download shared binaries.", default = false, type = "boolean"})
         add_configs("vs_runtime", {description = "Set vs compiler runtime.", default = "MT"})
     end
+
+    if is_plat("mingw") then
+        set_urls("https://github.com/xmake-mirror/build-artifacts/releases/download/mysql-$(version)/windows-x64-vc143-ad611e3230894c35896b2ccae5be9c7e.7z")
```
Comment (waruqi) [inline packages/m/mysql/xmake.lua:21]: 擦。。你直接搞得 build artifacts 哈，骚操作哈，不能这么玩。。这里merge 了，又会触发 build-artifacts ci，然后重新打包，hash 就变了。。

Comment (heheda123123) [reply packages/m/mysql/xmake.lua:21]: 我开始测试的时候是传到自己仓库的，
https://github.com/heheda123123/mysql-8.0.31
但是考虑可能存在信任安全问题，所以才在xmake-mirror那边提的一个pr
那这种咋处理呢？

Comment (heheda123123) [reply packages/m/mysql/xmake.lua:21]: 主要是mysql源码本身对mingw支持不好，改了好几次都编译不过，反正二进制兼容，就直接用msvc编译的了

File: packages/m/mysql/xmake.lua [hunk 3]
```diff
@@ -16,6 +16,13 @@ package("mysql")
         add_configs("shared", {description = "Download shared binaries.", default = false, type = "boolean"})
```
Comment (waruqi) [inline packages/m/mysql/xmake.lua:16]: 这里也不对，默认就是 static 。可以删了

File: packages/m/mysql/xmake.lua [hunk 4]
```diff
@@ -16,6 +16,13 @@ package("mysql")
...
         add_configs("vs_runtime", {description = "Set vs compiler runtime.", default = "MT"})
```
Comment (waruqi) [inline packages/m/mysql/xmake.lua:17]: 这里也删了，默认就是 MT

File: packages/m/mysql/xmake.lua (hunk 2 above)
Comment (waruqi) [reply packages/m/mysql/xmake.lua:21]: 可以上传到 mirror 回头我传下。。

**Comment by waruqi** [comment]
另外，每行尾部的空格去掉。。

image

File: packages/m/mysql/xmake.lua (hunk 2 above)
Comment (waruqi) [reply packages/m/mysql/xmake.lua:21]: 晚上给你传，https://github.com/xmake-mirror/mysql

现在没时间

**Comment by heheda123123** [comment]
好的，还有啥问题没，我再一起改了，不然每次一改就触发重新编译测试，会不会被github拉黑呀

**Comment by waruqi** [comment]
build-artifacts 上 x86 的都编译失败了。。只有 x64 的库 https://github.com/xmake-mirror/build-artifacts/actions/runs/4776706754/jobs/8491898939

很奇怪，之前怎么 merge 的。。这里 ci 没报出来。。

> \utilities\comp_client_err.vcxproj]
2023-04-23T06:02:56.3385435Z          mysys.lib(my_kdf.obj) : error LNK2019: unresolved external symbol _EVP_sha512 referenced in function "public: virtual int __thiscall Key_hkdf_function::derive_key(unsigned char const *,unsigned int,unsigned char *,unsigned int)" (?derive_key@Key_hkdf_function@@UAEHPBEIPAEI@Z) [C:\Users\runneradmin\AppData\Local\.xmake\cache\packages\2304\m\mysql\8.0.31\source\build_57a24fb2\utilities\comp_client_err.vcxproj]
2023-04-23T06:02:56.3386909Z          mysys.lib(my_kdf.obj) : error LNK2019: unresolved external symbol _PKCS5_PBKDF2_HMAC referenced in function "public: virtual int __thiscall Key_pbkdf2_hmac_function::derive_key(unsigned char const *,unsigned int,unsigned char *,unsigned int)" (?derive_key@Key_pbkdf2_hmac_function@@UAEHPBEIPAEI@Z) [C:\Users\runneradmin\AppData\Local\.xmake\cache\packages\2304\m\mysql\8.0.31\source\build_57a24fb2\utilities\comp_client_err.vcxproj]
2023-04-23T06:02:56.3388134Z          mysys.lib(my_kdf.obj) : error LNK2019: unresolved external symbol _EVP_PKEY_CTX_new_id referenced in function "public: virtual int __thiscall Key_hkdf_function::derive_key(unsigned char const *,unsigned int,unsigned char *,unsigned int)" (?derive_key@Key_hkdf_function@@UAEHPBEIPAEI@Z) [C:\Users\runneradmin\AppData\Local\.xmake\cache\packages\2304\m\mysql\8.0.31\source\build_57a24fb2\utilities\comp_client_err.vcxproj]
2023-04-23T06:02:56.3389755Z          mysys.lib(my_kdf.obj) : error LNK2019: unresolved external symbol _EVP_PKEY_CTX_free referenced in function "public: virtual int __thiscall Key_hkdf_function::derive_key(unsigned char const *,unsigned int,unsigned char *,unsigned int)" (?derive_key@Key_hkdf_function@@UAEHPBEIPAEI@Z) [C:\Users\runneradmin\AppData\Local\.xmake\cache\packages\2304\m\mysql\8.0.31\source\build_57a24fb2\utilities\comp_client_err.vcxproj]
2023-04-23T06:02:56.3391068Z          mysys.lib(my_kdf.obj) : error LNK2019: unresolved external symbol _EVP_PKEY_CTX_ctrl referenced in function "public: virtual int __thiscall Key_hkdf_function::derive_key(unsigned char const *,unsigned int,unsigned char *,unsigned int)" (?derive_key@Key_hkdf_function@@UAEHPBEIPAEI@Z) [C:\Users\runneradmin\AppData\Local\.xmake\cache\packages\2304\m\mysql\8.0.31\source\build_57a24fb2\utilities\comp_client_err.vcxproj]
2023-04-23T06:02:56.3392333Z          mysys.lib(my_kdf.obj) : error LNK2019: unresolved external symbol _EVP_PKEY_derive_init referenced in function "public: virtual int __thiscall Key_hkdf_function::derive_key(unsigned char const *,unsigned int,unsigned char *,unsigned int)" (?derive_key@Key_hkdf_function@@UAEHPBEIPAEI@Z) [C:\Users\runneradmin\AppData\Local\.xmake\cache\packages\2304\m\mysql\8.0.31\source\build_57a24fb2\utilities\comp_client_err.vcxproj]
2023-04-23T06:02:56.3393483Z          mysys.lib(my_kdf.obj) : error LNK2019: unresolved external symbol _EVP_PKEY_derive referenced in function "public: virtual int __thiscall Key_hkdf_function::derive_key(unsigned char const *,unsigned int,unsigned char *,unsigned int)" (?derive_key@Key_hkdf_function@@UAEHPBEIPAEI@Z) [C:\Users\runneradmin\AppData\Local\.xmake\cache\packages\2304\m\mysql\8.0.31\source\build_57a24fb2\utilities\comp_client_err.vcxproj]

**Comment by waruqi** [comment]
我把你加到 https://github.com/xmake-mirror/mysql 了。。不过不要把 binary 直接 push 进git，试试在 release 对应版本下，上传文件附件。。例如 mysql-windows.zip

**Comment by heheda123123** [comment]
> build-artifacts 上 x86 的都编译失败了。。只有 x64 的库 https://github.com/xmake-mirror/build-artifacts/actions/runs/4776706754/jobs/8491898939
> 
> 很奇怪，之前怎么 merge 的。。这里 ci 没报出来。。
> 
> > \utilities\comp_client_err.vcxproj]
> > 2023-04-23T06:02:56.3385435Z          mysys.lib(my_kdf.obj) : error LNK2019: unresolved external symbol _EVP_sha512 referenced in function "public: virtual int __thiscall Key_hkdf_function::derive_key(unsigned char const *,unsigned int,unsigned char *,unsigned int)" (?derive_key@Key_hkdf_function@@UAEHPBEIPAEI@Z) [C:\Users\runneradmin\AppData\Local.xmake\cache\packages\2304\m\mysql\8.0.31\source\build_57a24fb2\utilities\comp_client_err.vcxproj]
> > 2023-04-23T06:02:56.3386909Z          mysys.lib(my_kdf.obj) : error LNK2019: unresolved external symbol _PKCS5_PBKDF2_HMAC referenced in function "public: virtual int __thiscall Key_pbkdf2_hmac_function::derive_key(unsigned char const *,unsigned int,unsigned char *,unsigned int)" (?derive_key@Key_pbkdf2_hmac_function@@UAEHPBEIPAEI@Z) [C:\Users\runneradmin\AppData\Local.xmake\cache\packages\2304\m\mysql\8.0.31\source\build_57a24fb2\utilities\comp_client_err.vcxproj]
> > 2023-04-23T06:02:56.3388134Z          mysys.lib(my_kdf.obj) : error LNK2019: unresolved external symbol _EVP_PKEY_CTX_new_id referenced in function "public: virtual int __thiscall Key_hkdf_function::derive_key(unsigned char const *,unsigned int,unsigned char *,unsigned int)" (?derive_key@Key_hkdf_function@@UAEHPBEIPAEI@Z) [C:\Users\runneradmin\AppData\Local.xmake\cache\packages\2304\m\mysql\8.0.31\source\build_57a24fb2\utilities\comp_client_err.vcxproj]
> > 2023-04-23T06:02:56.3389755Z          mysys.lib(my_kdf.obj) : error LNK2019: unresolved external symbol _EVP_PKEY_CTX_free referenced in function "public: virtual int __thiscall Key_hkdf_function::derive_key(unsigned char const *,unsigned int,unsigned char *,unsigned int)" (?derive_key@Key_hkdf_function@@UAEHPBEIPAEI@Z) [C:\Users\runneradmin\AppData\Local.xmake\cache\packages\2304\m\mysql\8.0.31\source\build_57a24fb2\utilities\comp_client_err.vcxproj]
> > 2023-04-23T06:02:56.3391068Z          mysys.lib(my_kdf.obj) : error LNK2019: unresolved external symbol _EVP_PKEY_CTX_ctrl referenced in function "public: virtual int __thiscall Key_hkdf_function::derive_key(unsigned char const *,unsigned int,unsigned char *,unsigned int)" (?derive_key@Key_hkdf_function@@UAEHPBEIPAEI@Z) [C:\Users\runneradmin\AppData\Local.xmake\cache\packages\2304\m\mysql\8.0.31\source\build_57a24fb2\utilities\comp_client_err.vcxproj]
> > 2023-04-23T06:02:56.3392333Z          mysys.lib(my_kdf.obj) : error LNK2019: unresolved external symbol _EVP_PKEY_derive_init referenced in function "public: virtual int __thiscall Key_hkdf_function::derive_key(unsigned char const *,unsigned int,unsigned char *,unsigned int)" (?derive_key@Key_hkdf_function@@UAEHPBEIPAEI@Z) [C:\Users\runneradmin\AppData\Local.xmake\cache\packages\2304\m\mysql\8.0.31\source\build_57a24fb2\utilities\comp_client_err.vcxproj]
> > 2023-04-23T06:02:56.3393483Z          mysys.lib(my_kdf.obj) : error LNK2019: unresolved external symbol _EVP_PKEY_derive referenced in function "public: virtual int __thiscall Key_hkdf_function::derive_key(unsigned char const *,unsigned int,unsigned char *,unsigned int)" (?derive_key@Key_hkdf_function@@UAEHPBEIPAEI@Z) [C:\Users\runneradmin\AppData\Local.xmake\cache\packages\2304\m\mysql\8.0.31\source\build_57a24fb2\utilities\comp_client_err.vcxproj]

因为这次扔上去的libmysql.dll都是64位的，估计还要编译一套32位的dll才能通过check

**Comment by waruqi** [comment]
不是这个意思，是你上次 pr ci过了，但是 buuld-artifacts 那边 x86编译还是失败，没有 x86 dll 可用

**Pull request merged**